daffodil-runtime1/src/main/java/org/apache/daffodil/udf/UDFunctionService.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.udf;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.ServiceConfigurationError;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.ServiceLoader;
+import java.util.stream.Collectors;
+
+import org.apache.daffodil.util.Misc;
+
+public class UDFunctionService {
+	private ServiceLoader<UDFunctionProvider> loader;
+	private ArrayList<Class<?>> functionClasses = new ArrayList<>();
+	private HashMap<String, UDFunctionProvider> functionProviderLookup = new HashMap<>();
+	private ArrayList<String> errors = new ArrayList<>();
+	private ArrayList<String> warnings = new ArrayList<>();
+
+	public UDFunctionService(ClassLoader cl) {
+		try {
+			loader = cl == null ? ServiceLoader.load(UDFunctionProvider.class) : ServiceLoader.load(UDFunctionProvider.class, cl);
+
+			loader.forEach(fcp -> {
+				Class<?>[] fcpfc = fcp.getFunctionClasses();
+
+				if (fcpfc == null || fcpfc.length == 0) {
+					warnings.add(String.format("Provider ignored: [%s].\nNo Function Classes found.",
```

Comment:
Try to avoid \n in warning messsages. Same with the bracket. They just make things harder to read.

These arnings should probably also mention User Define Function Provider instead of just Provider. THe context might not be obvious to a user otherwise.

Reply:
Is it possible to pass in a ThrowsSDE context to create SDW/SDE instead of saving these warnings/errors to tmp vars and then displaying them later? Seems fragile if something ever forgets to read these and display them/act appropriately.

Reply:
As per our conversation, decided to use log(LogLevel.{Error,Info,Warning}...) where appropriate and then call SDE from the Expressions class if the (UserDefinedFunction, EvaluateMethod, ParameterTypes, ReturnType) tuple's not returned. Means we ran into an issue of some sort, either during loading or looking up the function.